strawberryfields/parameters.py

```diff
@@ -19,335 +19,334 @@
-  For now, we simply don't merge Operations that depend on RegRefTransforms or TensorFlow objects.
+  Measured and free parameters are handled symbolically by Sympy.
 
 * The compiled Program is run by a :class:`.BaseEngine` instance, which calls the
-  :meth:`~ops.Operation.apply` method of each Operation in turn.
+  :meth:`apply` method of each Operation in turn.
 
-* :meth:`Operation.apply` evaluates the numeric value of any RegRefTransform-based Parameters
-  using :meth:`Parameter.evaluate` (other types of Parameters are simply passed through).
-  The parameter values and the subsystem indices are passed to :meth:`Operation._apply`.
-
-* :meth:`~ops.Operation._apply` "unwraps" the Parameter instances. There are three different cases:
-
-  1. We still need to do some arithmetic, unwrap after it is done using `p.x`.
-  2. No arithmetic required, use :func:`~parameters._unwrap`.
-  3. No parameters are used, do nothing.
-
-  Finally, :meth:`_apply` calls the appropriate backend API method using the unwrapped parameters.
+* :meth:`apply` then calls :meth:`_apply` which is unique for each Operation subclass.
+  It may perform additional numeric or symbolic transformations on the parameters, and
+  then finally evaluates the numeric value of the parameters using :func:`par_evaluate`.
+  The numeric parameter values are passed to the appropriate backend API method.
   It is up to the backend to either accept NumPy arrays and Tensorflow objects as parameters, or not.
 
+
 What we cannot do at the moment:
 
-* Use anything except integers and RegRefs (or Sequences thereof) as the subsystem parameter
+* Use anything except integers and RegRefs (or Sequences thereof) as the subsystem argument
   for the :meth:`~ops.Operation.__or__` method.
-  Technically we could allow any Parameters or valid Parameter initializers that evaluate into an integer.
-* Do arithmetic with RegRefTransforms.
+  Technically we could allow any parameters that evaluate into an integer.
 
-Parameter methods
+
+Functions
+---------
+
+.. currentmodule:: strawberryfields.parameters
+
+.. autosummary::
+   par_evaluate
+   par_is_symbolic
+   par_regref_deps
+   par_str
+
+
+Parameter classes
 -----------------
 
-.. currentmodule:: strawberryfields.parameters.Parameter
+.. autosummary::
+   MeasuredParameter
+   FreeParameter
+
+
+Exceptions
+----------
 
 .. autosummary::
-   evaluate
+   ParameterError
 
 
 Code details
 ~~~~~~~~~~~~
 
 """
-import numbers
+# pylint: disable=too-many-ancestors,unused-argument,protected-access
+
+from collections.abc import Sequence
+import functools
+import types
 
-from unittest import mock
 import numpy as np
+import sympy
+import sympy.functions as sf
+
+
+def wrap_mathfunc(func):
+    """Applies the wrapped sympy function elementwise to numpy arrays.
+
+    Required because the sympy math functions cannot deal with arrays.
+    We implement no broadcasting; if the first argument is a numpy array, we assume
```

Comment:
Any possible edge-case bugs you can foresee?

Reply:
presumably `zip` would just give an error?

Reply:
Yeah. Most of the math funcs only take a single argument anyway.

Reply:
Actually zip will not give an error, it'll just stop when the first iterable is exhausted. I added some checks to fix this.